### PR TITLE
Add a single-threaded stdlib mode, use it for the 'minimal' stdlib

### PR DESCRIPTION
**Pull request by kubamracek**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -390,6 +390,10 @@ option(SWIFT_STDLIB_USE_NONATOMIC_RC
     "Build the standard libraries and overlays with nonatomic reference count operations enabled"
     FALSE)
 
+option(SWIFT_STDLIB_SINGLETHREADED_RUNTIME
+    "Build the standard libraries assuming that they will be used in an environment with only a single thread. If set, also forces SWIFT_STDLIB_SINGLETHREADED_RUNTIME to On."
```
[the remark this answers is not in the document]
Comment (kubamracek) [reply CMakeLists.txt]: Yes, I actually meant "SWIFT_STDLIB_USE_NONATOMIC_RC".

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 2]
```diff
@@ -329,6 +329,14 @@ function(_add_target_variant_c_compile_flags)
     list(APPEND result "-D_WASI_EMULATED_MMAN")
   endif()
 
+  if(SWIFT_STDLIB_USE_NONATOMIC_RC OR SWIFT_STDLIB_SINGLETHREADED_RUNTIME)
```
[the remark this answers is not in the document]
Comment (kubamracek) [reply stdlib/cmake/modules/AddSwiftStdlib.cmake]: Done.

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by gottesmm** [comment]
@kubamracek Are you sure that all rr operations will be non-atomic? I wouldn't rely on the current support in the compiler for emitting non-atomic rr. The compiler only attempts to keep things non-atomic as a "best effort". Better in your case to just add a flag to the runtime to ensure that ARC operations always go to the non-atomic variant.

**Comment by gottesmm** [comment]
@kubamracek or you could leave everything alone and just have them forward to the non-atomic implementation if the flag is set that way we aren't changing the external ABI interface of the library.

**Comment by kubamracek** [comment]
> Are you sure that all rr operations will be non-atomic?
No, and actually I'm pretty sure some will stay as atomic. However, my motivation here is just an "optimization", and if I discover places that are still using atomics when they don't need to, I plan to fix them up with separate PRs.

> add a flag to the runtime to ensure that ARC operations always go to the non-atomic variant
> just have them forward to the non-atomic implementation if the flag is set
That sounds reasonable. Let me try to prepare this in a separate PR.

**Comment by kubamracek** [comment]
@swift-ci please test

File: include/swift/Runtime/Mutex.h [hunk 3]
```diff
@@ -20,12 +20,14 @@
 
 #include <type_traits>
 
-#if (defined(__APPLE__) || defined(__linux__) || defined(__CYGWIN__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__HAIKU__))
+#ifdef SWIFT_STDLIB_SINGLETHREADED_RUNTIME
+#include "swift/Runtime/MutexSingleThreaded.h"
+#elif (defined(__APPLE__) || defined(__linux__) || defined(__CYGWIN__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__HAIKU__))
```
Comment (compnerd) [inline include/swift/Runtime/Mutex.h:25]: Not your doing, but I think that using `_POSIX_THREADS` is a better check for this.

File: stdlib/CMakeLists.txt [hunk 4]
```diff
@@ -17,6 +17,18 @@ option(SWIFT_ENABLE_COMPATIBILITY_OVERRIDES
   "Support back-deploying compatibility fixes for newer apps running on older runtimes."
   TRUE)
 
+option(SWIFT_STDLIB_USE_NONATOMIC_RC
+    "Build the standard libraries and overlays with nonatomic reference count operations enabled"
+    FALSE)
+
+option(SWIFT_STDLIB_SINGLETHREADED_RUNTIME
```
Comment (compnerd) [inline stdlib/CMakeLists.txt:24]: WDYT of spelling this `SWIFT_STDLIB_SINGLE_THREADED_RUNTIME`?  It makes it easier to read IMO.

File: stdlib/CMakeLists.txt [hunk 5]
```diff
@@ -17,6 +17,18 @@ option(SWIFT_ENABLE_COMPATIBILITY_OVERRIDES
...
+    "Build the standard libraries assuming that they will be used in an environment with only a single thread. If set, also forces SWIFT_STDLIB_USE_NONATOMIC_RC to On."
+    FALSE)
+
+if(SWIFT_STDLIB_SINGLETHREADED_RUNTIME)
+  set(SWIFT_STDLIB_USE_NONATOMIC_RC TRUE)
+endif()
```
Comment (compnerd) [inline stdlib/CMakeLists.txt:30]: I'm not suggesting that you change this, simply pointing out that there exists an alternative which requires that the user enable the option which is enforced via `cmake_dependent_option`.  What you have currently is fine, I'm just trying to get others better versed in CMake.

**Review by compnerd — APPROVED** [review]
The change looks reasonable to me.  I really like the idea that @gottesmm has to use shims, and if the shims are static I think that could be an interesting approach as well.  Since this form hasn't shipped yet it does make sense to simply do the more straightforward changes that @kubamracek has proposed.

LGTM, but please make sure that @gottesmm is okay with it as well before merging.

File: include/swift/Runtime/Mutex.h (hunk 3 above)
Comment (kubamracek) [reply include/swift/Runtime/Mutex.h:25]: https://github.com/apple/swift/pull/33609

File: stdlib/CMakeLists.txt (hunk 4 above)
Comment (kubamracek) [reply stdlib/CMakeLists.txt]: Yup. Switched to that new spelling.

File: stdlib/CMakeLists.txt (hunk 5 above)
Comment (kubamracek) [reply stdlib/CMakeLists.txt:34]: Good to know! I'll keep this as is, if you're okay with that, and keep `cmake_dependent_option` in mind.

**Comment by kubamracek** [comment]
@gottesmm Are you okay with the state of this PR?

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by gottesmm** [comment]
I talked with Kuba offline. I asked him in follow on 1-2 follow on PRs to:

1. Eliminate NONATOMIC_RC. That was just an experimental flag from some time ago that isn't needed today.
2.  Change the 10 atomic operations in HeapObject.cpp to be non-atomic if the Single threaded runtime mode is set. The change is really small. Example:

```
HeapObject *swift::swift_retain(HeapObject *object) {
  CALL_IMPL(swift_retain, (object));
}
```

becomes

```
HeapObject *swift::swift_retain(HeapObject *object) {
#if SINGLE_THREAD_FLAG
  CALL_IMPL(swift_nonatomic_retain, (object));
#else
  CALL_IMPL(swift_retain, (object));
#endif
}
```

Thanks for your work here Kuba!

**Comment by benrimmington** [comment]
@swift-ci Please test Windows platform

**Pull request merged**